haystack/pipelines/base.py

```diff
@@ -1159,69 +1160,79 @@ def get_config(self, return_defaults: bool = False) -> dict:
+            component: BaseComponent = node_attributes["component"]
+            if node_name != component.name:
+                raise PipelineError(f"Component name '{component.name}' does not match node name '{node_name}'.")
+
+            self._add_component_to_definitions(
+                component=component, component_definitions=component_definitions, return_defaults=return_defaults
+            )
 
             # create the Pipeline definition with how the Component are connected
-            pipelines[pipeline_name]["nodes"].append({"name": node, "inputs": list(self.graph.predecessors(node))})
+            pipeline_definitions[pipeline_name]["nodes"].append(
+                {"name": node_name, "inputs": list(self.graph.predecessors(node_name))}
+            )
 
         config = {
-            "components": list(components.values()),
-            "pipelines": list(pipelines.values()),
+            "components": list(component_definitions.values()),
+            "pipelines": list(pipeline_definitions.values()),
             "version": __version__,
         }
         return config
 
-    def _generate_component_name(self, type_name: str, params: Dict[str, Any], existing_components: Dict[str, Any]):
+    def _add_component_to_definitions(
+        self, component: BaseComponent, component_definitions: Dict[str, Dict], return_defaults: bool = False
+    ):
+        if component.name is None:
+            raise PipelineError(f"Component with config '{component._component_config}' does not have a name.")
+
+        component_params: Dict[str, Any] = component.get_params(return_defaults)
+        # handling of subcomponents: add to definitions and substitute by reference
+        for param_key, param_value in component_params.items():
+            if isinstance(param_value, BaseComponent):
+                sub_component = param_value
+                self._add_component_to_definitions(sub_component, component_definitions, return_defaults)
+                component_params[param_key] = sub_component.name
+
+        component_definitions[component.name] = {
+            "name": component.name,
+            "type": component.type,
+            "params": component_params,
+        }
+
+    def _get_all_component_names(self, components_to_search: Optional[List[BaseComponent]] = None) -> Set[str]:
+        component_names = set()
+        if components_to_search is None:
+            components_to_search = list(self.components.values())
+        for component in components_to_search:
+            if component.name is not None:
+                component_names.add(component.name)
+                sub_component_names = self._get_all_component_names(component.sub_components)
+                component_names.update(sub_component_names)
+        return component_names
+
+    def _set_sub_component_names(self, component: BaseComponent, component_names: Optional[Set[str]] = None):
```

Comment:
Unfortunately during `_add_component_to_definitions()` we do not have all the component_names already assigned within that pipeline. That's a requirement for generating names in order to not end up with duplicates. So `_get_all_component_names` would be required either way.
I also thought about assigning (virtual) names during `get_config()` and thus `in _add_component_to_definitions()`. However it occured cleaner to me that all the names are already being assigned when the component is added to the pipeline because that already happens for non-sub-components.